### PR TITLE
chore!: convert project to typescript

### LINKED ISSUE #1640: FEATURE - Convert to Typescript

### PR DESCRIPTION
**Pull request by robertsLando**
Fixes #1640

### CODE REVIEW

**Review by sibelius — APPROVED** [review]
awesome

**Comment by robertsLando** [comment]
@vishnureddy17  @BertKleewein All tests are working now 🚀

File: src/lib/connect/index.ts
Comment (vishnureddy17) [inline src/lib/connect/index.ts:56]: maybe do function overloading here? one signature with just an opts param and another signature with the url string and opts

**Comment by vishnureddy17** [comment]
Is it ready for review or are you still working on it? @robertsLando

**Comment by robertsLando** [comment]
@vishnureddy17 the code is ready for review, I was just testing that `npm pack` produces a valid package with all exports correctly set.

File: src/lib/TypedEmitter.ts
Comment (vishnureddy17) [inline src/lib/TypedEmitter.ts:4]: Why do we need to have these types? Shouldn't there be typings available for EventEmitter already?

**Review by vishnureddy17 — APPROVED** [review]
I've skimmed through and made some comments. It is a big PR, so I did not go through each line very thoroughly. Let me know if there are particular parts of the PR that you think should be look at more closely.

File: src/lib/TypedEmitter.ts
Comment (robertsLando) [reply src/lib/TypedEmitter.ts:4]: A friend of mine (@AlCalzone, a TS wizard 🧙🏼‍♂️ ) suggested me this a while ago. Based on what he told me the `TypedEventEmitter` is far better because:

> you get autocomplete for the events that exist, and you get syntax help for the callback parameters, and typescript will yell at you if you accidentally use an event wrong. No more typos in event names or non-existent events, no more wrong callback params, no more looking up elsewhere what callback params there are

File: src/lib/connect/index.ts
Comment (robertsLando) [reply src/lib/connect/index.ts]: @vishnureddy17 I have been able to add them, check: [`1c56921` (#1641)](https://github.com/mqttjs/MQTT.js/pull/1641/commits/1c569211bf6aaa961e13b89f877d4677ae64278a)

File: src/lib/connect/ws.ts
Comment (robertsLando) [reply src/lib/connect/ws.ts]: fixed

File: src/lib/TypedEmitter.ts
Comment (AlCalzone) [reply src/lib/TypedEmitter.ts:4]: Examples for a class that is such a `TypedEventEmitter`:
grafik
grafik
grafik

**Comment by robertsLando** [comment]
@vishnureddy17 @BertKleewein I have end up here, I will check the code another time to see if I spot something but I would merge this if you agree. If you want to take some time to review this just let me know and I will wait till next week before merging

File: src/lib/TypedEmitter.ts
Comment (vishnureddy17) [reply src/lib/TypedEmitter.ts:4]: Ah, neat!

**Review by vishnureddy17 — APPROVED** [review]

**Comment by vishnureddy17** [comment]
If you think it's good to merge, go for it

**Pull request merged**
